### ISSUE TITLE
.gitignore folder negation not working correctly

### ISSUE THREAD

**Issue by cbargren** [libgit2/libgit2#3989]
Here's an example of what I mean —

.gitignore:
```
# ignore everything
*

# still search in folders though
!*/

# don’t ignore .c files
!*.c
```

Folder structure
```
foo.dll    # ignored
src
  bar.dll  # not ignored, but should be
  foo.c    # not ignored
```

Output of `git status -u`:
```
On branch master
Untracked files:
  (use "git add <file>..." to include in what will be committed)

	src/foo.c
```

Output of `git status --ignored -u`:
```
On branch master
Untracked files:
  (use "git add <file>..." to include in what will be committed)

	src/foo.c

Ignored files:
  (use "git add -f <file>..." to include in what will be committed)

	foo
	src/bar.dll
```

**Comment by cbargren** [comment]
I spent a decent amount of time today looking into this and at this point, I'm not really sure that the [.gitignore docs](https://git-scm.com/docs/gitignore) even define what this behavior should be. My vague understanding is that `*/` should ignore any directory and anything under any directory at the root level, so `!*/` should _include_ any directory and anything under a directory at the root level. That's not what the output of `git status -u` from above is indicating though. Am I misunderstanding anything or could this actually be an inconsistency in git?

**Comment by Asadarian** [comment]
It doesn't matter whether it is consistent with the docs. It works in the original git, and many large projects (for example (Unreal Engine) are using it, so this library must support it aswell.

**Comment by amaiorano** [comment]
Yes, please fix this so that it's consistent with Git. I'm trying to make use of libgit2sharp with Unreal Engine, and I end up with a massive list of "untracked" files, none of which show up with 'git status' as they should be ignored.

**Comment by swisspol** [comment]
Isn't this related to (or a dup of) #3224?

**Comment by pks-t** [comment]
@swisspol It probably is and will hopefully be fixed by #4296. Will investigate as soon as it is merged

**Comment by JSn1nj4** [comment]
Hey guys. Any recent updates on this issue?

**Comment by bloodelves88** [comment]
Hi, I would like to give this a little nudge, since certain repositories like Unreal Engine could really use to have this working. I currently have 17386 unstaged files which should all be ignored, but they aren't due to this bug. It's impossible to work in the repository using libgit.

I see the PR's CI test has failed. Any chance that it'll be fixed in the near future?

**Comment by pks-t** [comment]
The failures were just unrelated errors due to some previous problems with our CI setup. The PR has landed now, so it would be nice if you could test if current master still exhibits the same erroneous behaviour.

**Comment by gallexme** [comment]
Any Updates on this?

**Comment by ethomson** [comment]
> Any Updates on this?

Hi @gallexme - yes, the PR was merged.  Are you seeing problems still?  Can you describe them?

**Comment by gallexme** [comment]
According to the git kraken team yes:



Thank you for taking the time to reach out to support. The issue you have
described is a bug in one of the dependencies that we use for GitKraken. We
use libgit2 and use its .gitignore patterns. Here is the link to the bug
report https://github.com/libgit2/libgit2/issues/3989. You can ignore these
files directly though GitKrkan just keep in mind that this will update the
current .gitignore file.

I hope this information is helpful. Please let me know if you have any
questions.



Am Di., 13. Nov. 2018, 16:34 hat Edward Thomson <notifications@github.com>
geschrieben:

**Comment by pks-t** [comment]
The fix for #3989 has been merged via #4296, which is part of v0.26.3+ and v0.27.0+. I don't know which version of libgit2 Git Kraken is used, and I didn't get any feedback whether the changes I did fixed the issues that were seen. So from our point of view, we have fixed the faulty cases and so we don't intend to do anything additional.

If you or somebody from Git Kraken confirms though that they have the fixes from v0.26.3/v0.27.0 included already and they _still_ see issues, then I'd naturally be happy to fix them.
